Server/ReasnAPI/ReasnAPI/Controllers/EventsController.cs

```diff
@@ -1,171 +1,307 @@
     {
-        throw new NotImplementedException();
+        var relatedEvent = eventService.GetEventBySlug(slug);
+        var participating = eventService.GetEventParticipantsCountBySlugAndStatus(slug, ParticipantStatus.Participating);
+        var interested = eventService.GetEventParticipantsCountBySlugAndStatus(slug, ParticipantStatus.Interested);
+        var username = relatedEvent.Organizer.Username;
+        var participants = new Participants(participating, interested);
+        var images = eventService.GetEventImages(slug);
+        var eventResponse = relatedEvent.ToDto().ToResponse(participants, username, $"/api/v1/Users/image/{username}", relatedEvent.Address.ToDto(), relatedEvent.AddressId, images);
+
+        return Ok(eventResponse);
     }
 
     [HttpPut]
     [Authorize(Roles = "Admin, Organizer")]
     [Route("{slug}")]
-    public IActionResult UpdateEvent(string slug)
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    public IActionResult UpdateEvent([FromRoute] string slug, [FromBody] EventUpdateRequest eventUpdateRequest, [FromServices] IValidator<EventDto> validator)
     {
-        throw new NotImplementedException();
+        
+        var existingEvent = eventService.GetEventBySlug(slug);
+        var user = userService.GetCurrentUser();
+        var eventDto = eventUpdateRequest.ToDto(user.Id);
+        eventDto.Slug = slug;
+        validator.ValidateAndThrow(eventDto);
+        if (existingEvent.OrganizerId != user.Id && user.Role != UserRole.Admin)
+        {
+            return Forbid();
+        }
+
+        if (existingEvent.Status != eventUpdateRequest.Status && user.Role != UserRole.Admin && eventUpdateRequest.Status != EventStatus.Cancelled)
+        {
+            return Forbid();
+        }
+        eventService.UpdateAddressForEvent(eventUpdateRequest.AddressDto, existingEvent.AddressId, slug);
+        eventService.UpdateEvent(existingEvent.Id, eventDto);
+
+        return Ok();
     }
 
     [HttpGet]
     [Authorize(Roles = "Admin")]
     [Route("requests")]
-    public IActionResult GetEventsRequests(string slug)
+    [ProducesResponseType<List<EventResponse>>(StatusCodes.Status200OK)]
+    public IActionResult GetEventsRequests()
     {
-        throw new NotImplementedException();
+        var events = eventService.GetEventsByFilter(e => e.Status == EventStatus.PendingApproval);
+       
+        return Ok(events); 
     }
 
     [HttpPost]
     [Authorize(Roles = "Admin")]
     [Route("{slug}/approve")]
-    public IActionResult ApproveEventRequest(string slug)
+    [ProducesResponseType<EventDto>(StatusCodes.Status200OK)]
+    public IActionResult ApproveEventRequest([FromRoute] string slug)
     {
-        throw new NotImplementedException();
+        var eventToApprove = eventService.GetEventBySlug(slug);
+
+        eventToApprove.Status = EventStatus.Approved;
+        eventService.UpdateEvent(eventToApprove.Id, eventToApprove.ToDto());
+        return Ok();
+    }

```

Comment:
Do we have a decline one?

Reply:
i caleated one for receting eventrequest